File: javascript/webdriver/command.js
// Copyright 2011 Software Freedom Conservancy. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @fileoverview Contains several classes for handling commands.
 */

goog.provide('webdriver.Command');
goog.provide('webdriver.CommandExecutor');
goog.provide('webdriver.CommandName');
goog.provide('webdriver.CommandResponse');



/**
 * Describes a command to be executed by the WebDriverJS framework.
 * @param {!webdriver.CommandName} name The name of this command.
 * @constructor
 * @export
 */
webdriver.Command = function(name) {

  /**
   * The name of this command.
   * @type {!webdriver.CommandName}
   * @private
   */
  this.name_ = name;

  /**
   * The parameters to this command.
   * @type {!Object.<*>}
   * @private
   */
  this.parameters_ = {};
};


/**
 * @return {!webdriver.CommandName} This command's name.
 * @export
 */
webdriver.Command.prototype.getName = function() {
  return this.name_;
};


/**
 * Sets a parameter to send with this command.
 * @param {string} name The parameter name.
 * @param {*} value The parameter value.
 * @return {!webdriver.Command} A self reference.
 * @export
 */
webdriver.Command.prototype.setParameter = function(name, value) {
  this.parameters_[name] = value;
  return this;
};


/**
 * Sets the parameters for this command.
 * @param {!Object.<*>} parameters The command parameters.
 * @return {!webdriver.Command} A self reference.
 * @export
 */
webdriver.Command.prototype.setParameters = function(parameters) {
  this.parameters_ = parameters;
  return this;
};


/**
 * Returns a named command parameter.
 * @param {string} key The parameter key to look up.
 * @return {*} The parameter value, or undefined if it has not been set.
 * @export
 */
webdriver.Command.prototype.getParameter = function(key) {
  return this.parameters_[key];
};


/**
 * @return {Object.<*>} The parameters to send with this command.
 * @export
 */
webdriver.Command.prototype.getParameters = function() {
  return this.parameters_;
};


/**
 * Enumeration of predefined names command names that all command processors
 * will support. Implemented as a map so the enumeration can be properly
 * exported as part of WebDriver's public API.
 * @type {!Object.<string>}
 * @export
 */
// TODO(jleyba): Delete obsolete command names.
webdriver.CommandName = {
  'GET_SERVER_STATUS': 'status',

  'NEW_SESSION': 'newSession',
  'GET_SESSIONS': 'getSessions',
  'DESCRIBE_SESSION': 'getSessionCapabilities',

  'CLOSE': 'close',
  'QUIT': 'quit',

  'GET_CURRENT_URL': 'getCurrentUrl',
  'GET': 'get',
  'GO_BACK': 'goBack',
  'GO_FORWARD': 'goForward',
  'REFRESH': 'refresh',

  'ADD_COOKIE': 'addCookie',
  'GET_COOKIE': 'getCookie',
  'GET_ALL_COOKIES': 'getCookies',
  'DELETE_COOKIE': 'deleteCookie',
  'DELETE_ALL_COOKIES': 'deleteAllCookies',

  'GET_ACTIVE_ELEMENT': 'getActiveElement',
  'FIND_ELEMENT': 'findElement',
  'FIND_ELEMENTS': 'findElements',
  'FIND_CHILD_ELEMENT': 'findChildElement',
  'FIND_CHILD_ELEMENTS': 'findChildElements',

  'CLEAR_ELEMENT': 'clearElement',
  'CLICK_ELEMENT': 'clickElement',
  'SEND_KEYS_TO_ELEMENT': 'sendKeysToElement',
  'SUBMIT_ELEMENT': 'submitElement',
  'TOGGLE_ELEMENT': 'toggleElement',

  'GET_CURRENT_WINDOW_HANDLE': 'getCurrentWindowHandle',
  'GET_WINDOW_HANDLES': 'getWindowHandles',

  'SWITCH_TO_WINDOW': 'switchToWindow',
  'SWITCH_TO_FRAME': 'switchToFrame',
  'GET_PAGE_SOURCE': 'getPageSource',
  'GET_TITLE': 'getTitle',

  'EXECUTE_SCRIPT': 'executeScript',
  'EXECUTE_ASYNC_SCRIPT': 'executeAsyncScript',

  'GET_ELEMENT_TEXT': 'getElementText',
  'GET_ELEMENT_TAG_NAME': 'getElementTagName',
  'IS_ELEMENT_SELECTED': 'isElementSelected',
  'IS_ELEMENT_ENABLED': 'isElementEnabled',
  'IS_ELEMENT_DISPLAYED': 'isElementDisplayed',
  'GET_ELEMENT_LOCATION': 'getElementLocation',
  'GET_ELEMENT_SIZE': 'getElementSize',
  'GET_ELEMENT_ATTRIBUTE': 'getElementAttribute',
  'GET_ELEMENT_VALUE_OF_CSS_PROPERTY': 'getElementValueOfCssProperty',
  'ELEMENT_EQUALS': 'elementEquals',

  'SCREENSHOT': 'screenshot',
  'DIMISS_ALERT': 'dimissAlert',
  'IMPLICITLY_WAIT': 'implicitlyWait',
  'SET_SCRIPT_TIMEOUT': 'setScriptTimeout',

  'GET_ALERT': 'getAlert',
  'ACCEPT_ALERT': 'acceptAlert',
  'DISMISS_ALERT': 'dismissAlert',
  'GET_ALERT_TEXT': 'getAlertText',
  'SET_ALERT_VALUE': 'setAlertValue'
};


/**
 * Type definition for a WebDriver response object as defined by the wire
 * protocol.
 * @typedef {{status:bot.ErrorCode, value:*}}
 */
webdriver.CommandResponse;


/**
 * Handles the execution of {@code webdriver.Command} objects.
 * @interface
 */
webdriver.CommandExecutor = function() {};


/**
 * Executes the given {@code command}. If there is an error executing the
 * command, the provided callback will be invoked with the offending error.
 * Otherwise, the callback will be invoked with a null Error and non-null
 * {@code webdriver.CommandResponse} object.
 * @param {!webdriver.Command} command The command to execute.
 * @param {function(Error, !webdriver.CommandResponse=)} callback the function
 *     to invoke when the command response is ready.
 */
webdriver.CommandExecutor.prototype.execute = goog.abstractMethod;

File: dotnet/src/webdriver/Command.cs
// <copyright file="Command.cs" company="Selenium Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// </copyright>

using OpenQA.Selenium.Internal;
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;

namespace OpenQA.Selenium;

/// <summary>
/// Provides a way to send commands to the remote server
/// </summary>
public class Command
{
    private readonly static JsonSerializerOptions s_jsonSerializerOptions = new()
    {
        TypeInfoResolverChain =
        {
            CommandJsonSerializerContext.Default,
            new DefaultJsonTypeInfoResolver()
        },
        Converters = { new ResponseValueJsonConverter() }
    };

    /// <summary>
    /// Initializes a new instance of the <see cref="Command"/> class using a command name and a JSON-encoded string for the parameters.
    /// </summary>
    /// <param name="name">Name of the command</param>
    /// <param name="jsonParameters">Parameters for the command as a JSON-encoded string.</param>
    public Command(string name, string jsonParameters)
        : this(null, name, ConvertParametersFromJson(jsonParameters))
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="Command"/> class for a Session
    /// </summary>
    /// <param name="sessionId">Session ID the driver is using</param>
    /// <param name="name">Name of the command</param>
    /// <param name="parameters">Parameters for that command</param>
    /// <exception cref="ArgumentNullException">If <paramref name="name"/> is <see langword="null"/>.</exception>
    public Command(SessionId? sessionId, string name, Dictionary<string, object?>? parameters)
    {
        this.SessionId = sessionId;
        this.Parameters = parameters ?? new Dictionary<string, object?>();
        this.Name = name ?? throw new ArgumentNullException(nameof(name));
    }

    /// <summary>
    /// Gets the SessionID of the command
    /// </summary>
    [JsonPropertyName("sessionId")]
    public SessionId? SessionId { get; }

    /// <summary>
    /// Gets the command name
    /// </summary>
    [JsonPropertyName("name")]
    public string Name { get; }

    /// <summary>
    /// Gets the parameters of the command
    /// </summary>
    [JsonPropertyName("parameters")]
    public Dictionary<string, object?> Parameters { get; }

    /// <summary>
    /// Gets the parameters of the command as a JSON-encoded string.
    /// </summary>
    public string ParametersAsJsonString
    {
        get
        {
            if (this.Parameters != null && this.Parameters.Count > 0)
            {
                return JsonSerializer.Serialize(this.Parameters, s_jsonSerializerOptions);
            }
            else
            {
                return "{}";
            }
        }
    }

    /// <summary>
    /// Returns a string of the Command object
    /// </summary>
    /// <returns>A string representation of the Command Object</returns>
    public override string ToString()
    {
        return string.Concat("[", this.SessionId, "]: ", this.Name, " ", this.ParametersAsJsonString);
    }

    /// <summary>
    /// Gets the command parameters as a <see cref="Dictionary{K, V}"/>, with a string key, and an object value.
    /// </summary>
    /// <param name="value">The JSON-encoded string representing the command parameters.</param>
    /// <returns>A <see cref="Dictionary{K, V}"/> with a string keys, and an object value. </returns>
    /// <exception cref="JsonException">If <paramref name="value"/> is not a JSON object.</exception>
    /// <exception cref="ArgumentNullException">If <paramref name="value"/> is <see langword="null"/>.</exception>
    private static Dictionary<string, object?>? ConvertParametersFromJson(string value)
    {
        Dictionary<string, object?>? parameters = JsonSerializer.Deserialize<Dictionary<string, object?>>(value, CommandJsonSerializerContext.Default.DictionaryStringObject!);
        return parameters;
    }
}

// Built-in types
[JsonSerializable(typeof(bool))]
[JsonSerializable(typeof(byte))]
[JsonSerializable(typeof(sbyte))]
[JsonSerializable(typeof(char))]
[JsonSerializable(typeof(decimal))]
[JsonSerializable(typeof(double))]
[JsonSerializable(typeof(float))]
[JsonSerializable(typeof(int))]
[JsonSerializable(typeof(uint))]
[JsonSerializable(typeof(nint))]
[JsonSerializable(typeof(nuint))]
[JsonSerializable(typeof(long))]
[JsonSerializable(typeof(ulong))]
[JsonSerializable(typeof(short))]
[JsonSerializable(typeof(ushort))]
[JsonSerializable(typeof(string))]

// Selenium WebDriver types
[JsonSerializable(typeof(char[]))]
[JsonSerializable(typeof(byte[]))]
[JsonSerializable(typeof(Chromium.ChromiumNetworkConditions))]
[JsonSerializable(typeof(Cookie))]
[JsonSerializable(typeof(ReturnedCookie))]
[JsonSerializable(typeof(Proxy))]

// Selenium Dictionaries, primarily used in Capabilities
[JsonSerializable(typeof(Dictionary<string, object>))]
[JsonSerializable(typeof(Dictionary<string, bool>))]
[JsonSerializable(typeof(Dictionary<string, byte>))]
[JsonSerializable(typeof(Dictionary<string, sbyte>))]
[JsonSerializable(typeof(Dictionary<string, char>))]
[JsonSerializable(typeof(Dictionary<string, decimal>))]
[JsonSerializable(typeof(Dictionary<string, double>))]
[JsonSerializable(typeof(Dictionary<string, float>))]
[JsonSerializable(typeof(Dictionary<string, int>))]
[JsonSerializable(typeof(Dictionary<string, uint>))]
[JsonSerializable(typeof(Dictionary<string, nint>))]
[JsonSerializable(typeof(Dictionary<string, nuint>))]
[JsonSerializable(typeof(Dictionary<string, long>))]
[JsonSerializable(typeof(Dictionary<string, ulong>))]
[JsonSerializable(typeof(Dictionary<string, short>))]
[JsonSerializable(typeof(Dictionary<string, ushort>))]
[JsonSerializable(typeof(Dictionary<string, string>))]
[JsonSourceGenerationOptions(Converters = [typeof(ResponseValueJsonConverter)])]
internal partial class CommandJsonSerializerContext : JsonSerializerContext;
